src/countsHelper.cpp

```diff
@@ -0,0 +1,31 @@
+#include <Rcpp.h>
+#include <unordered_map>
+#include <vector>
+using namespace Rcpp;
+
+//' @title
+//' Count number of occurences
+//'
+//' @param features a Matrix with doubles.
+//'
+//' @export
+// [[Rcpp::export]]
+NumericMatrix  countsToDhist(NumericMatrix features)
+{
+    std::unordered_map<double,int> counter;
+    int i;
+    // Silly solution to the float comparison problem
+    for (i=0;i<features.nrow();i++)
+    {counter[features[i]]=0;}
+    for (i=0;i<features.nrow();i++)
+    {counter[features[i]]+=1;}
+    std::vector<std::pair<double,int> > keys(counter.begin(),counter.end());
```

Comment:
I'm not quite sure what this syntax does. I can't quite map it to one of the initialisers at http://en.cppreference.com/w/cpp/container/vector/vector

Reply:
Looking at what we do with it later, I'm assuming it creates a std::vector from the map, somehow assigning the map key to the first element in each pair and the map value to the second element?

Reply:
I am using the following pattern from stackoverflow
https://stackoverflow.com/questions/31323135/sort-an-unordered-map-using-sort

Basically the counter.begin() and counter.end() creates an iterator over the map and it is using the range vector creation method from:

http://www.cplusplus.com/reference/vector/vector/vector/

There are other patterns that can be used (indeed this pattern is not in this list):

https://stackoverflow.com/questions/19588420/how-to-transform-map-into-vector-using-transform-function-template

There may be an argument for using a map rather than an unordered_map, as we are going to order later, but as the lookup average complexity is O(log(k)) (it is often a tree), whereas unordered_map is O(1). Thus the overall average complexity with n items with k unique values should be O(n log(k)) for the map, but O(n+klog(k)) for the unordered map I thought that this made more sense, but it is possible that the constants kill this argument for the range of values we care about....

Reply:
Ah, yes. I see the creation method in my original list now. I missed that the third parameter had a default value.